improvement(client): formalize IFluidContainerInternal

jason-ha opened the pull request:
as internal only interface with `container` accessible for devtools and experimental access.

Simplify devtools with strict expectation that `IFluidContainer` is implemented by FF.

Additionally, remove `inheritDoc` use in non-exported class FluidContainer to make IntelliSense work under VS Code.

jason-ha committed:
improvement(client): formalize IFluidContainerInternal
with INTERNAL_CONTAINER_DO_NOT_USE

jason-ha committed:
docs(client): remove inheritDoc
from non-exported class
as VS Code does not support. VS Code will link docs without inheritDoc from interfaces implemented; with inheritDoc there is nothing but "inheritDoc ..."

CraigMacomber commented on packages/tools/devtools/devtools/api-report/devtools.alpha.api.md:
```diff
@@ -32,6 +32,4 @@ export { IDevtoolsLogger }
 // @beta
 export function initializeDevtools(props: DevtoolsProps): IDevtools;
 
-// (No @packageDocumentation comment for this package)
```
this delta seems. odd. There was a package documentation comment, so this shouldn't exist, but why was it here before your change? Did somone not update the API report and CI not catch that somehow?

Josmithr replied:
But the devtools package _does_ have a `@packageDocumentation` comment.
https://github.com/microsoft/FluidFramework/blob/main/packages/tools/devtools/devtools/src/index.ts#L18

jason-ha replied:
Exactly - don't know why the report said it didn't prior to this change.

Josmithr replied:
🤦‍♂️ Parsed the diff backwards.

Josmithr commented on packages/framework/fluid-static/src/fluidContainer.ts:
```diff
@@ -230,6 +230,20 @@ export interface IFluidContainer<TContainerSchema extends ContainerSchema = Cont
+	 *
+	 * @remarks Used to power debug tooling and experimental features.
+	 */
+	INTERNAL_CONTAINER_DO_NOT_USE?(): IContainer;
```
Since this is internal only, we should probably rename it now. And since this was only optional for pre-internal backwards compat, we should make this required.
```suggestion
	readonly container: IContainer;
```

jason-ha replied:
I think this one is a little sneaky.
`IFluidContainer` is now `@sealed` which I think means we can assume that if we get an `IFluidContainer` then we expect it to come from use and thus have `IFluidContainerInternal`. So, if we had an alternate implementation that didn't also support `IFluidContainerInternal` that would be on us fully.
Do we all agree on that thinking? (+ @CraigMacomber )

If we keep it as is, then we are free to have `IFluidContainer` that does not have `IFluidContainerInternal`. The current pattern is very safe as the method is optional and forces the user to do a check even after casting.

If we change, I guess we do something like introduce:
```typescript
function toInternalFluidContainer<TContainerSchema extends ContainerSchema>(
	container: IFluidContainer<TContainerSchema> | IFluidContainerInternal
): IFluidContainerInternal {
	if ("INTERNAL_CONTAINER_DO_NOT_USE" in container) {
		return container;
	}
	throw new Error("container is not IFluidContainerInternal");
}
```
But does devtools want to be that strict? Currently it logs console error and moves on.

So, perhaps:
```typescript
/**
 * Internal interface for {@link IFluidContainer}.
 * @internal
 */
export interface IFluidContainerInternal {
	/**
	 * The underlying {@link @fluidframework/container-definitions#IContainer}.
	 * @remarks Used to power debug tooling and experimental features.
	 */
	readonly container: IContainer;
}

/**
 * Check that the provided `container` is an internal {@link IFluidContainerInternal}.
 * @internal
 */
export function isInternalFluidContainer<TContainerSchema extends ContainerSchema>(
	container: IFluidContainer<TContainerSchema> | IFluidContainerInternal,
): container is IFluidContainerInternal {
	return "container" in container;
}
```
and devtools would look like:
```typescript
function mapContainerProps(
	containerProps: ContainerDevtoolsProps,
): ContainerDevtoolsPropsBase | undefined {
	const { container, containerKey } = containerProps;
	if (!isInternalFluidContainer(container)) {
		console.error("Missing Container accessor on FluidContainer.");
		return undefined;
	}

	return {
		container: container.container,
		containerKey,
		containerData: container.initialObjects as Record<string, IFluidLoadable>,
	};
}
```

Josmithr replied:
I think it's fine for devtools to be strict about the presence of the internal container. If users want to consumer our devtools, then they must follow our documented contracts and not introduce their own `IFluidContainer` implementations.

jason-ha replied:
Updated to failing faster - left the internal utility still "relaxed"; caller decides how to handle

jason-ha committed:
improvement(client): fully leverage internal types
simplify devtools with expectation that only `FluidContainer` is expected/allowed implementation of `IFluidContainer`.

Josmithr commented on packages/framework/fluid-static/src/fluidContainer.ts:
```diff
@@ -245,7 +259,21 @@ export function createFluidContainer<
+ *
+ * @internal
+ */
+export function isInternalFluidContainer<TContainerSchema extends ContainerSchema>(
```
I wonder if it might be better (or at least easier) to just export the `FluidContainer` class as `@internal`, and make the `container` property public.

This pattern also seems fine, but also seems unnecessary since this is for internal use only.

jason-ha replied:
In general, I think we should hide classes, even internal ones. `/internal` doesn't mean inaccessible even if no one outside should. When there isn't a class exposed, things are just that much safer/stable. Easier to expose it eventually if we need that than to walk-it back.
(I see Bohemia using /internal things even though they shouldn't.)

Josmithr replied:
Yeah, that's fine. And we can easily change this in the future if we want to.

Josmithr commented on packages/tools/devtools/devtools/src/index.ts:
```diff
@@ -173,18 +162,14 @@ export function initializeDevtools(props: DevtoolsProps): IDevtools {
-		console.error("Missing Container accessor on FluidContainer.");
-		return undefined;
+	if (!isInternalFluidContainer(container)) {
+		throw new TypeError("Container is not required FluidContainer.");
```
Since theoretically end-users will see this, we should be a bit more descriptive here. Note that the `IFluidContainer` implementation was unrecognized, and only Containers generated by the Fluid-Framework are supported.

jason-ha committed:
improvement(client): update user visible error text
